internal/provisioner/kops_provisioner_cluster_installation_beta.go

```diff
@@ -414,6 +423,91 @@ func (provisioner *kopsCIBeta) getMattermostCustomResource(cluster *model.Cluste
 	return cr, nil
 }
 
+func (provisioner *kopsCIBeta) migrateFromClusterInstallation(clusterInstallation *model.ClusterInstallation, k8sClient *k8s.KubeClient, logger log.FieldLogger) error {
+	name := makeClusterInstallationName(clusterInstallation)
+	ciClient := k8sClient.MattermostClientsetV1Alpha.MattermostV1alpha1().ClusterInstallations(clusterInstallation.Namespace)
+	mmClient := k8sClient.MattermostClientsetV1Beta.MattermostV1beta1().Mattermosts(clusterInstallation.Namespace)
+
+	ctx := context.TODO()
+	cr, err := ciClient.Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		if k8sErrors.IsNotFound(err) {
+			return nil
+		}
+
+		return errors.Wrapf(err, "failed to get cluster installation %s", clusterInstallation.ID)
+	}
+
+	if cr.Spec.Migrate {
+		if cr.Status.Migration.Error != "" {
+			err = errors.Errorf("error while migrating ClusterInstallation: %s", cr.Status.Migration.Error)
+			logger.Error(err)
+			return err
+		}
+
+		logger.Infof("Migration already in progress, status: %s", cr.Status.Migration.Status)
+		return provisioner.waitForMigration(name, mmClient, logger)
+	}
+
+	logger.Infof("Starting CR migration to Mattermost")
+
+	cr.Spec.Migrate = true
+	cr, err = ciClient.Update(ctx, cr, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to start migration of the ClusterInstallation")
+	}
+
+	return provisioner.waitForMigration(name, mmClient, logger)
+}
+
+func (provisioner *kopsCIBeta) waitForMigration(name string, mmClient v1beta1.MattermostInterface, logger log.FieldLogger) error {
```

Comment:
Unfortunately, I think this method of migration may not work well for us. Despite the logic of the migration working perfectly, the wait loop contained here will tie up the ClusterInstallation Supervisor during this period. Because the supervisors all tick and run sequentially this basically ties up the provisioner entirely from processing other work during this time.

I believe we will need to alter this to perform a check and to quickly exit and continue on to other resources if the migration is still underway.

I know that we use similar wait loops elsewhere like in some cluster flows, but the reason we can get away with it there is simply because we separate out our cluster supervisors to only run on certain provisioners and we also don't make cluster changes all that often.

Reply:
Yes, I was on the fence on that. I was thinking that running separate instances will give us enough space for other things as we will probably want to run migration slowly anyway, to not to kill databases. Also the migration is one time thing and the alternative implementation that I had in mind was kind of messy as it would introduce more potential states.

I see your point though, and after rethinking that I came to similar conclusion. As you said it would be better to just do the check.

Reply:
I also thought about running separate migration-only provisioners to slowly migrate, but unfortunately due to how work is picked up by provisioners randomly I could see issues with ensuring all installations get migrated. That said, your new commit looks good in that it will allow us to let every provisioner do migrations now. Also, good job doing the that without new installation states.

Regarding overloading databases, I was also concerned about that, but I think we are okay here because we can allow the group rolling update behavior to ensure that we don't go too quickly when we do the migrations.